source/slang/slang-ir-variable-scope-correction.cpp

```diff
@@ -0,0 +1,307 @@
+            // variable visible in the whole function.
+            UInt operandCount = newInst->getOperandCount();
+            for (UInt i = 0; i < operandCount; i++)
+            {
+                auto operand = newInst->getOperand(i);
+
+                // workList contains all the instructions in the loop region, so if the operands in the list, then we have to
+                // handle the operand as well
+                if (workList.contains(operand))
+                {
+                    // Recursively check all the operands
+                    // Now the newly inserted inst is the user of the operand, so we pass 'operand' as the origin instruction, and
+                    // 'newInst' as the user instruction.
+                    if (auto newOperand = _processInstruction(dominatorTree, instAfterParam, operand, newInst, loopHeaderList, workList))
+                    {
+                        // Replace the operand with the new instruction
+                        newInst->setOperand(i, newOperand);
+                    }
+                }
+            }
+        }
+        // Replace the operands of user with the new instruction
+        _replaceOperand(useInst, originInst, newInst);
+        return newInst;
+    }
+    return nullptr;
+}
+
+// 1. Duplicate the original instruction
+// 2. Insert the duplicated instruction right before the use site
+// 3. return the load instruction
+IRInst* VariableScopeCorrectionContext::_processUnstorableInst(IRInst* inst, IRInst* user)
+{
+    IRCloneEnv cloneEnv;
+    m_builder.setInsertBefore(user);
+
+    // duplicate the invisible instruction and insert it right before the use site
+    auto clonedInst = cloneInst(&cloneEnv, &m_builder, inst);
+    clonedInst->insertAt(IRInsertLoc::before(user));
+
+    return clonedInst;
+}
+
+// 1. Declare a new variable at the beginning of the function
+// 2. Insert a store instruction after the original instruction
+// 3. Insert a load instruction before the use site
+// 4. Return the load instruction
+IRInst* VariableScopeCorrectionContext::_processStorableInst(IRInst* insertLoc, IRInst* inst, IRInst* user)
+{
+    auto type = inst->getDataType();
+    // store instruction must have a result type
+    SLANG_ASSERT(type);
+
+    // declare a new variable at the beginning of the function used to store the result of the instruction
+    m_builder.setInsertBefore(insertLoc);
+    auto dstPtr = m_builder.emitVar(type);
+
+    // insert a store instruction after the instruction
+    m_builder.setInsertAfter(inst);
+    m_builder.emitStore(dstPtr, inst);
+
+    // last, replace operands in the use site instruction with the new variable
+    // Note, because "dstPtr" is a pointer type, we have to insert a load(dstPtr) instruction before use it.
+    // Simply replace any operand with pointer could generate error code.
+    m_builder.setInsertBefore(user);
```

Comment:
Instead of handling just one user here, simply loop through all uses and insert loads at all out of scope use sites.